server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -210,15 +215,26 @@ SafeCommitInfo getSafeCommitInfo() {
      * @param acquiringSafeCommit captures the most recent safe commit point if true; otherwise captures the most recent commit point.
      */
     synchronized IndexCommit acquireIndexCommit(boolean acquiringSafeCommit) {
+        return acquireIndexCommit(acquiringSafeCommit, false);
+    }
+
+    private synchronized IndexCommit acquireIndexCommit(boolean acquiringSafeCommit, boolean acquiredInternally) {
         assert safeCommit != null : "Safe commit is not initialized yet";
         assert lastCommit != null : "Last commit is not initialized yet";
         final IndexCommit snapshotting = acquiringSafeCommit ? safeCommit : lastCommit;
         acquiredIndexCommits.merge(snapshotting, 1, Integer::sum); // increase refCount
-        return wrapCommit(snapshotting);
+        if (Assertions.ENABLED && acquiredInternally) {
+            internallyAcquiredIndexCommits.add(snapshotting);
+        }
```

Comment:
We should check `add` return true to ensure no commit is added more than once. I also think we can simplify this with

```suggestion
        assert acquiredInternally == false || internallyAcquiredIndexCommits.add(snapshotting) : "commit [" + snapshotting + "] already added";
```